### PR TITLE
fix: WASM provided by nix build of barretenberg

### PR DESCRIPTION
**Pull request by kobyhallx**
# Related issue(s)
    
#23     
# Resolves # 

# Description

Switching Noir to barretenberg upstream introduced requirement for Barretenberg WASM to be located at `$BARRETENBERG_BIN_DIR`. Absence of this variable breaks builds.
    
## Summary of changes

libbarretenberg-wasm32 is built by nix from upstream AztecProtocol/barretenberg, uploaded to artifacts and later downloaded and reused by workflows.    
    
## Dependency additions / changes

libbarretenberg-wasm32 
    
## Test additions / changes
N/A    

# Checklist
    
- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

## Documentation needs
- [ ] This PR requires documentation updates when merged.
    
# Additional context

### CODE REVIEW

File: .github/workflows/publish-apple-darwin-wasm.yml [hunk 1]
```diff
@@ -11,8 +20,36 @@ on:
...
+      - uses: cachix/cachix-action@v12
+        with:
+          name: barretenberg
+          authToken: "${{ secrets.CACHIX_AUTH_TOKEN }}"
+
+      # Upload does not work with symlinks, using this workaround:
+      # https://github.com/actions/upload-artifact/issues/92#issuecomment-1080347032    
+      - name: Build barretenberg as libbarretenberg-wasm32
+        run: |
+          nix build "git+https://github.com/AztecProtocol/barretenberg#wasm32"
```
Comment (phated) [inline .github/workflows/publish-apple-darwin-wasm.yml]: This won't be pinned to the bb commit that noir is pinned to. We need to be using the lockfile that noir is

File: .github/workflows/publish-apple-darwin-wasm.yml [hunk 2]
```diff
@@ -11,8 +20,36 @@ on:
...
+          echo "ARTIFACT_UPLOAD_PATH=$(readlink -f result)" >> $GITHUB_ENV
+
+      - name: Upload artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: libbarretenberg-wasm32
+          path: ${{ env.ARTIFACT_UPLOAD_PATH }}
+          retention-days: 3
```
Comment (phated) [inline .github/workflows/publish-apple-darwin-wasm.yml]: What's the reason for making this a separate job and then uploading the artifact? Could we combine into one job and not upload?

**Review by phated — CHANGES_REQUESTED** [review]
This is a great solution. We need to figure out how to pin bb to the commit noir uses though, perhaps using the lockfile?

File: .github/workflows/publish-apple-darwin-wasm.yml (hunk 2 above)
Comment (kobyhallx) [reply .github/workflows/publish-apple-darwin-wasm.yml]: in separate update to be converted to workflow_run invocation to share one build across all platform workflows

File: .github/workflows/publish-apple-darwin-wasm.yml (hunk 1 above)
Comment (kobyhallx) [reply .github/workflows/publish-apple-darwin-wasm.yml]: which lockfile?

Comment (phated) [reply .github/workflows/publish-apple-darwin-wasm.yml]: https://github.com/noir-lang/noir/blob/master/flake.lock#L15-L17

File: .github/workflows/publish-apple-darwin-wasm.yml (hunk 2 above)
Comment (phated) [reply .github/workflows/publish-apple-darwin-wasm.yml]: Isn't the wasm pre-built in the cachix binary cache though? I'd expect us to just be downloading the wasm file

Comment (kobyhallx) [reply .github/workflows/publish-apple-darwin-wasm.yml]: it's not only that, windows won't run nix and needs to download artifcat from somewhere anyway

File: .github/workflows/publish-apple-darwin-wasm.yml (hunk 1 above)
Comment (kobyhallx) [reply .github/workflows/publish-apple-darwin-wasm.yml]: pinning done in nix-barretenberg.yml

File: .github/workflows/nix-barretenberg.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      uses: actions/checkout@v3
+      with:
+        repository: noir-lang/noir
+        path: noir
+        ref: ${{ inputs.noir-ref || 'master' }}
+
+    - name: Collect locked barretenberg rev
+      run: |
+        echo "BB_REV=$(jq -r .nodes.barretenberg.locked.rev ./noir/flake.lock)" >> $GITHUB_ENV
+
```
Comment (TomAFrench) [inline .github/workflows/nix-barretenberg.yml]:
```suggestion
```

nit

File: .github/workflows/publish-apple-darwin-wasm.yml [hunk 4]
```diff
@@ -1,14 +1,12 @@
...
-  workflow_dispatch:
-    inputs:
-      noir-ref:
-        description: The noir reference to checkout
-        required: false
-  schedule:
-    - cron: "0 2 * * *" # run at 2 AM UTC
-  push:
+on: 
+  workflow_run:
```
Comment (phated) [inline .github/workflows/publish-apple-darwin-wasm.yml:4]: This doesn't trigger a workflow when another runs, it is used to make re-usable workflows that can be called as jobs.

Comment (kobyhallx) [reply .github/workflows/publish-apple-darwin-wasm.yml:4]: as specified line below this will get triggered by `libbarretenberg` workflow completion

Comment (phated) [reply .github/workflows/publish-apple-darwin-wasm.yml:4]: ~~No, that's not how these work~~

I was incorrect, these only work on default branch, which @kobyhallx forked and test. See action runs at https://github.com/kobyhallx/build-nargo-nl/actions

File: .github/workflows/publish-apple-darwin-wasm.yml [hunk 5]
```diff
@@ -45,13 +43,21 @@ jobs:
             target/
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
+      - name: Download artifact
+        uses: dawidd6/action-download-artifact@v2
```
Comment (phated) [inline .github/workflows/publish-apple-darwin-wasm.yml:48]: Can we use a github official action or not? https://github.com/actions/download-artifact

**Review by phated — CHANGES_REQUESTED** [review]
Question about your choice about an unofficial action when an official one exists. Maybe there is a reason we can't use official version?

File: .github/workflows/publish-apple-darwin-wasm.yml [hunk 6]
```diff
@@ -45,13 +43,21 @@ jobs:
             target/
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
+      - name: Download artifact
```
Comment (phated) [inline .github/workflows/publish-apple-darwin-wasm.yml:47]:
```suggestion
      # We use an unofficial download plugin to support cross-workflow downloads
      # Ref https://github.com/actions/download-artifact/issues/172
      - name: Download artifact
```

File: .github/workflows/publish-linux.yml [hunk 7]
```diff
@@ -48,13 +45,22 @@ jobs:
             target/
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
+      - name: Download pre-built libbarretenberg-wasm32
```
Comment (phated) [inline .github/workflows/publish-linux.yml]:
```suggestion
      # We use an unofficial download plugin to support cross-workflow downloads
      # Ref https://github.com/actions/download-artifact/issues/172
      - name: Download artifact
```

File: .github/workflows/publish-x86_64-pc-windows-wasm.yml [hunk 8]
```diff
@@ -40,13 +36,21 @@ jobs:
             target/
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
+      - name: Download artifact
```
Comment (phated) [inline .github/workflows/publish-x86_64-pc-windows-wasm.yml:41]:
```suggestion
      # We use an unofficial download plugin to support cross-workflow downloads
      # Ref https://github.com/actions/download-artifact/issues/172
      - name: Download artifact
```

**Review by phated — CHANGES_REQUESTED** [review]
Let's add this comment to each workflow.

File: .github/workflows/publish-apple-darwin-wasm.yml (hunk 5 above)
Comment (kobyhallx) [reply .github/workflows/publish-apple-darwin-wasm.yml:48]: we tried, but official one does not work cross workflows as seen on 
https://github.com/actions/download-artifact/issues/172

**Review by phated — APPROVED** [review]
We need to update Noir to dispatch to this new workflow.

**Pull request merged**